src/threading.c

```diff
@@ -905,16 +924,180 @@ void _jl_mutex_unlock_nogc(jl_mutex_t *lock)
 #endif
 }
 
-void _jl_mutex_unlock(jl_task_t *self, jl_mutex_t *lock)
+JL_DLLEXPORT void _jl_spin_mutex_unlock(jl_task_t *self, jl_spin_mutex_t *lock)
 {
-    _jl_mutex_unlock_nogc(lock);
+    _jl_spin_mutex_unlock_nogc(lock);
     jl_lock_frame_pop(self);
     JL_SIGATOMIC_END_self();
     if (jl_atomic_load_relaxed(&jl_gc_have_pending_finalizers)) {
         jl_gc_run_pending_finalizers(self); // may GC
     }
 }
 
+JL_DLLEXPORT void _jl_sleep_mutex_init(jl_sleep_mutex_t *lock, const char *name) JL_NOTSAFEPOINT
+{
+    // waiters is overloaded:
+    // bit 31 indicates sleep lock
+    // bits 30:1 indicate number of waiter threads
+    // bit 0 indicates lock held
+    lock->count = 0;
+    jl_atomic_store_relaxed(&lock->owner, (jl_task_t*)NULL);
+    jl_atomic_store_release(&lock->waiters, SLEEP_LOCK_BIT);
+    jl_profile_lock_init(lock, name);
+}
+
+JL_DLLEXPORT void _jl_sleep_mutex_wait(jl_task_t *self, jl_sleep_mutex_t *lock, int safepoint)
+{
+    // high bit of waiters identifies this as a sleep lock
+    // call out the uninitialized case specially
+    assert(jl_atomic_load_relaxed(&lock->waiters) != 0 && "Sleep lock not initialized!");
+    assert(!is_spin_mutex(lock) && "Unexpected spin lock in sleeping wait!");
+    uint32_t old_waiters = SLEEP_LOCK_BIT;
+    if (jl_atomic_cmpswap_acqrel(&lock->waiters, &old_waiters, SLEEP_LOCK_BIT | 1)) {
+        // no one is waiting, we just took the lock
+        jl_atomic_store_relaxed(&lock->owner, self);
+        lock->count = 1;
+        return;
+    }
+    if (jl_atomic_load_relaxed(&lock->owner) == self) {
+        // we already own the lock, just increment the count
+        lock->count++;
+        return;
+    }
+    jl_profile_lock_start_wait(lock);
+    //Do some retries?
+    jl_atomic_fetch_add(&lock->waiters, 0b10);
+    int8_t gc_state;
+    if (safepoint)
+        gc_state = jl_gc_safe_enter(self->ptls);
+    uintptr_t hash = (uintptr_t)lock;
+    hash >>= SLEEP_IGNORED_BITS;
```

Comment:
How safe is this? Do these hashes really not need to be unique?

Reply:
I would worry a bit about highly unpredictable performance phenomena due to this.

Reply:
There is a predictable number of threads, so even one sleep_lock would be fine, but more just helps avoid some percentage of spurious wakeups